File: sdk/ruby/lib/rockbox/api/playlist.rb
# frozen_string_literal: true

require_relative "../types"

module Rockbox
  module Api
    class Playlist
      TRACK_FIELDS = <<~GQL
        fragment TrackFields on Track {
          id title artist album genre disc trackString yearString
          composer comment albumArtist grouping
          discnum tracknum layer year bitrate frequency
          filesize length elapsed path
          albumId artistId genreId albumArt
        }
      GQL

      def initialize(http)
        @http = http
      end

      # @return [Rockbox::Playlist]
      def current
        data = @http.execute(<<~GQL)
          #{TRACK_FIELDS}
          query CurrentPlaylist {
            playlistGetCurrent {
              amount index maxPlaylistSize firstIndex
              lastInsertPos seed lastShuffledStart
              tracks { ...TrackFields }
            }
          }
        GQL
        pl = data[:playlist_get_current] || {}
        Rockbox::Playlist.new(
          amount:              pl[:amount],
          index:               pl[:index],
          max_playlist_size:   pl[:max_playlist_size],
          first_index:         pl[:first_index],
          last_insert_pos:     pl[:last_insert_pos],
          seed:                pl[:seed],
          last_shuffled_start: pl[:last_shuffled_start],
          tracks:              Array(pl[:tracks]).map { |t| Track.from_hash(t) }
        )
      end

      def amount
        @http.execute("query PlaylistAmount { playlistAmount }")[:playlist_amount]
      end

      # ---------------------------------------------------------------------
      # Queue management
      # ---------------------------------------------------------------------

      # @param paths       [Array<String>] file paths or track IDs to insert.
      # @param position    [Integer] one of {Rockbox::InsertPosition} (default: NEXT).
      # @param playlist_id [String, nil] target playlist; nil for the active queue.
      def insert_tracks(paths, position: InsertPosition::NEXT, playlist_id: nil)
        @http.execute(
          "mutation InsertTracks($playlistId: String, $position: Int!, $tracks: [String!]!) { " \
          "insertTracks(playlistId: $playlistId, position: $position, tracks: $tracks) }",
          { playlist_id: playlist_id, position: position, tracks: paths }
        )
        nil
      end

      def insert_directory(directory, position: InsertPosition::LAST, playlist_id: nil)
        @http.execute(
          "mutation InsertDirectory($playlistId: String, $position: Int!, $directory: String!) { " \
          "insertDirectory(playlistId: $playlistId, position: $position, directory: $directory) }",
          { playlist_id: playlist_id, position: position, directory: directory }
        )
        nil
      end

      def insert_album(album_id, position: InsertPosition::LAST)
        @http.execute(
          "mutation InsertAlbum($albumId: String!, $position: Int!) { " \
          "insertAlbum(albumId: $albumId, position: $position) }",
          { album_id: album_id, position: position }
        )
        nil
      end

      def remove_track(index)
        @http.execute(
          "mutation RemoveTrack($index: Int!) { playlistRemoveTrack(index: $index) }",
          { index: index }
        )
        nil
      end

      def clear
        @http.execute("mutation ClearPlaylist { playlistRemoveAllTracks }")
        nil
      end

      def shuffle
        @http.execute("mutation ShufflePlaylist { shufflePlaylist }")
        nil
      end

      def create(name, tracks)
        @http.execute(
          "mutation CreatePlaylist($name: String!, $tracks: [String!]!) { " \
          "playlistCreate(name: $name, tracks: $tracks) }",
          { name: name, tracks: tracks }
        )
        nil
      end

      def start(start_index: nil, elapsed: nil, offset: nil)
        @http.execute(
          "mutation PlaylistStart($startIndex: Int, $elapsed: Int, $offset: Int) { " \
          "playlistStart(startIndex: $startIndex, elapsed: $elapsed, offset: $offset) }",
          { start_index: start_index, elapsed: elapsed, offset: offset }.compact
        )
        nil
      end

      def resume
        @http.execute("mutation PlaylistResume { playlistResume }")
        nil
      end
    end
  end
end

File: sdk/clojure/src/rockbox/playlist.clj
(ns rockbox.playlist
  "Live playback queue management — what's currently playing and what's
  queued up next. For persistent named playlists see `rockbox.saved-playlists`."
  (:refer-clojure :exclude [shuffle])
  (:require [rockbox.transport :as t]
            [rockbox.types     :as types]))

(def ^:private track-fields
  "fragment TrackFields on Track {
     id title artist album genre disc trackString yearString
     composer comment albumArtist grouping
     discnum tracknum layer year bitrate frequency
     filesize length elapsed path
     albumId artistId genreId albumArt
   }")

(defn current
  "The active queue: `:tracks :amount :index :max-playlist-size ...`."
  [client]
  (:playlist-get-current
   (t/execute client
              (str track-fields
                   " query CurrentPlaylist {
                       playlistGetCurrent {
                         amount index maxPlaylistSize firstIndex
                         lastInsertPos seed lastShuffledStart
                         tracks { ...TrackFields }
                       }
                     }"))))

(defn amount
  "Number of tracks in the active queue."
  [client]
  (:playlist-amount (t/execute client "query PlaylistAmount { playlistAmount }")))

;; ---------------------------------------------------------------------------
;; Queue management
;; ---------------------------------------------------------------------------

(defn insert-tracks
  "Insert track paths (or IDs) into the queue.

  `position` is a `rockbox.types/insert-position` keyword (`:next :after-current
  :last :first`) or the underlying integer; defaults to `:next`.

  Optional `:playlist-id` targets a specific playlist instead of the active queue."
  ([client paths] (insert-tracks client paths :next nil))
  ([client paths position] (insert-tracks client paths position nil))
  ([client paths position playlist-id]
   (t/execute client
              "mutation InsertTracks($playlistId: String, $position: Int!, $tracks: [String!]!) {
                 insertTracks(playlistId: $playlistId, position: $position, tracks: $tracks)
               }"
              {:playlist-id playlist-id
               :position    (types/->insert-position position)
               :tracks      (vec paths)})
   client))

(defn insert-directory
  "Insert a directory's contents (recursively) into the queue."
  ([client directory] (insert-directory client directory :last nil))
  ([client directory position] (insert-directory client directory position nil))
  ([client directory position playlist-id]
   (t/execute client
              "mutation InsertDirectory($playlistId: String, $position: Int!, $directory: String!) {
                 insertDirectory(playlistId: $playlistId, position: $position, directory: $directory)
               }"
              {:playlist-id playlist-id
               :position    (types/->insert-position position)
               :directory   directory})
   client))

(defn insert-album
  "Append all tracks from an album to the queue."
  ([client album-id] (insert-album client album-id :last))
  ([client album-id position]
   (t/execute client
              "mutation InsertAlbum($albumId: String!, $position: Int!) {
                 insertAlbum(albumId: $albumId, position: $position)
               }"
              {:album-id album-id :position (types/->insert-position position)})
   client))

(defn remove-track
  "Remove the track at queue index `i` (0-based)."
  [client i]
  (t/execute client "mutation RemoveTrack($index: Int!) { playlistRemoveTrack(index: $index) }"
             {:index i})
  client)

(defn clear
  "Remove every track from the queue."
  [client]
  (t/execute client "mutation ClearPlaylist { playlistRemoveAllTracks }")
  client)

(defn shuffle
  "Shuffle the remaining tracks in the queue."
  [client]
  (t/execute client "mutation ShufflePlaylist { shufflePlaylist }")
  client)

(defn create
  "Create and start a new temporary queue from a list of paths.
  Replaces the current queue."
  [client name paths]
  (t/execute client
             "mutation CreatePlaylist($name: String!, $tracks: [String!]!) {
                playlistCreate(name: $name, tracks: $tracks)
              }"
             {:name name :tracks (vec paths)})
  client)

(defn start
  "Start playback of the current queue. Options:
   `:start-index` (int), `:elapsed` (int), `:offset` (int)."
  ([client] (start client {}))
  ([client opts]
   (t/execute client
              "mutation PlaylistStart($startIndex: Int, $elapsed: Int, $offset: Int) {
                 playlistStart(startIndex: $startIndex, elapsed: $elapsed, offset: $offset)
               }"
              opts)
   client))

(defn resume
  "Resume from the saved position."
  [client]
  (t/execute client "mutation PlaylistResume { playlistResume }")
  client)
